### PR TITLE
KEYCLOAK-4544 Broker Login autolink with disabled user creation

### PR DESCRIPTION
**Pull request by diodfr**
Hi,
I am unable to reopen the PR #6282 .
Like other users describe in the JIRA ticket, I use an Id provider and I want to restrict users by linking only on existing account.
According to @mposolda comments, I have added a DISABLE_USER_CREATION option to the IdpCreateUserIfUniqueAuthenticator.

Hope my contribution will help.
Didier

### CODE REVIEW

**Comment by Hokwang** [comment]
Hi, this will be resolve my problem(https://issues.redhat.com/browse/KEYCLOAK-15629) too, right?

**Comment by mposolda** [comment]
@diodfr Thanks for the PR and sorry for late review. I will try to take a look at your PR in the next few days.

**Comment by Hokwang** [comment]
@mposolda waiting your review

**Comment by mposolda** [comment]
@diodfr I am sorry for the late response. When looking at this PR, there is one confusing thing which is the fact that "Create User If Unique" authenticator doesn't create user at all even if it stated this in it's name.

So instead of adding the config option to this authenticator, I vote for creating new authenticator implementation like "Detect existing broker user". It will be ideal if it either inherits from the "Create User If Unique" authenticator or have the same subclass with the "Create User If Unique" authenticator to avoid as much code duplications as possible and just move the stuff where both implementation differs to some protected methods (or something along the lines of this). WDYT?

File: services/src/main/java/org/keycloak/authentication/AuthenticationProcessor.java [hunk 1]
```diff
@@ -976,12 +976,12 @@ public Response authenticateOnly() throws AuthenticationFlowException {
         AuthenticationFlow authenticationFlow = createFlowExecution(this.flowId, null);
         Response challenge = authenticationFlow.processFlow();
         if (challenge != null) return challenge;
-        if (authenticationSession.getAuthenticatedUser() == null) {
```
Comment (mposolda) [inline services/src/main/java/org/keycloak/authentication/AuthenticationProcessor.java:979]: Is it possible to avoid changes in this class in this PR?

**Review by mposolda — CHANGES_REQUESTED** [review]
Added the comment above and also one inline comment in the AuthenticatorProcessor class.

**Comment by diodfr** [comment]
> @diodfr I am sorry for the late response. When looking at this PR, there is one confusing thing which is the fact that "Create User If Unique" authenticator doesn't create user at all even if it stated this in it's name.
> 
> So instead of adding the config option to this authenticator, I vote for creating new authenticator implementation like "Detect existing broker user". It will be ideal if it either inherits from the "Create User If Unique" authenticator or have the same subclass with the "Create User If Unique" authenticator to avoid as much code duplications as possible and just move the stuff where both implementation differs to some protected methods (or something along the lines of this). WDYT?

Hi @mposolda,
That is what I have tried to do 1 year ago on my previous PR :(.
Ok I will try to take time to do it again.
Thanks for your review.

**Comment by jasonchester** [comment]
For some additional context, https://github.com/keycloak/keycloak/pull/6282 is the previous PR from @diodfr which introduced a different authenticator and was rejected. 

@mposolda I understand the difficult tasks a reviewer has maintaining the quality of the code base, this is needed functionality for a good portion of the keycloak user base and I am sure many would appreciate it being included.

My personal opinion is that a single authenticator with a configurable option is a better user experience for the user, makes documentation simpler and configuration more discoverable than having multiple new authenticators.

I might suggest a name like **Handle Existing IDP User** for this combined functionality.

**Comment by tulequ** [comment]
currently I use following snippet to temporary solve this problem
https://gist.github.com/tulequ/f0d810d7bfcf6cb5db3bba5aad07b7cd

We need setup flow as
```
1. User Must Exists --> ALTERNATIVE
2. Automatically Set Existing User --> ALTERNATIVE
```

This code is based on "Create User If Unique"

Hope this help!!!

**Comment by mposolda** [comment]
@diodfr Thanks! Good luck. I hope that we finally managed to have this in.

**Comment by mposolda** [comment]
@tulequ Thanks for pointing your solution.

**Comment by Hokwang** [comment]
@diodfr Hi, happy new year! Do you have a plan for update this issue?

**Comment by diodfr** [comment]
@Hokwang Hi, happy new year for you to ! I will try to work on it this week.

**Comment by mposolda** [comment]
@diodfr Thanks a lot!

**Comment by Hokwang** [comment]
@diodfr sorry and thanks, I am waiting your update.

**Comment by diodfr** [comment]
@mposolda hope this time it will be better :) I will do the documentation once you are agree to merge the PR.
Once again thanks for your work. I like this tool.
Regards
Didier

**Review by mposolda — CHANGES_REQUESTED** [review]
@diodfr Thanks for the updated PR!

It looks great. I've only added 2 minor comments inline.

The missing part is documentation. I think that the PR might be needed to https://github.com/keycloak/keycloak-documentation/ . I think it will be nice to update "First Broker Login" flow chapter of the documentation and add some example with this authenticator used instead of "Create User If Unique" authenticator. Will be good to describe the difference between the setup with this authenticator and the setup used in this section https://www.keycloak.org/docs/latest/server_admin/index.html#_disabling_automatic_user_creation . Maybe that section can be updated to contain 2 examples (one example with your authenticator and the second, which already exists there)?

**Comment by diodfr** [comment]
@mposolda I have modified the PR according your comments.
The documentation is available here : https://github.com/keycloak/keycloak-documentation/pull/1103

Thanks for your review. Let me know if there is anything wrong.
Regards,
Didier

**Comment by mposolda** [comment]
@diodfr I am sorry, but I see in your test that during error scenario, user will still see the error message "Invalid username or password." . This error message does not makes much sense for this error use-case and it is likely confusing for users. Perhaps something better like `User authenticated with identity provider 'facebook' does not exists. Please contact your administrator.` or something like this?

I pointed the error Messages.INVALID_USER before, but I did not check that this error message is mapped to "Invalid username or password." . It is my bad and I am sorry for the confusion. Will it be possible if you add new error message similar to what I specified above? You may need to add new constant to the class `Messages` and also new corresponding sentence for this message key into file `themes/src/main/resources/theme/base/login/messages/messages_en.properties`. In the authenticator class, you can use something like this to map the alias of IDP provider to the message `setErrorMessage(Messages.FEDERATED_IDENTITY_UNAVAILABLE, brokerContext.getIdpConfig().getAlias()`  or something like this. WDYT?

**Comment by diodfr** [comment]
@mposolda I have updated the error message according to your review comment.

**Review by mposolda — APPROVED** [review]
@diodfr Thanks! Re-running the tests. They failed, but it seems unrelated to your PR. CI is just sometimes not very stable :)

**Comment by syrenius** [comment]
Thanks for the contribution, looking for a solution of the decribed scenario brought me here...
Do I understand right, that the commit cb12fed96eeeffd118694856ffe7c453e33844a8 should be part of the release 12.0.4? 
If yes, the "Detect existing broker user Authenticater Execution" seems not to appear in the list for me:
image

**Comment by syrenius** [comment]
Thank you very much for the contribution, this feature/scenario would be very helpful for us!
I have noticed that despite the PR has been merged into the master branch before the release date, the latest release (12.0.4) does not yet contain the "Detect existing broker user" authenticator. I asume this is due to the tests failing?

As we build on the official docker image, I would be very thankful for a short estimate, whether it is likely that this authenticator will be integrated into the next release, or if it will take some more time and we rather should create our own image based on the current version of the master branch? Thanks!

**Comment by Hokwang** [comment]
@syrenius This PR will be included next major version release. for now, you need to build docker image if you want to test this PR.

**Comment by gsustek** [comment]
Hi, i tested and this feature (https://www.keycloak.org/docs/latest/server_admin/#automatically-link-existing-first-login-flow) and this feature(https://www.keycloak.org/docs/latest/server_admin/#_disabling_automatic_user_creation) on KC 13.0.0, with token_exchange:
1. obtain ext JWT with 
`curl -v --insecure --request POST \
  --url http://192.168.168.211:8081/auth/realms/ext-idp/protocol/openid-connect/token \
  --header 'content-type: application/x-www-form-urlencoded' \
  --header 'x-csrf-token: 1' \
  --data grant_type=password \
  --data username=test \
  --data password=test \
  --data client_id=oidc-client-from-intIdp \
  --data client_secret=3e5e4547-cd0d-49af-84cc-5c87981be7ef \
  --data scope='openid email profile' \
  | jq -r '.access_token'`
2. ext to int on intIdp
` curl --request POST \
  --url http://192.168.168.211:8080/auth/realms/int-idp/protocol/openid-connect/token \
  --header 'content-type: application/x-www-form-urlencoded' \
  --header 'x-csrf-token: 1' \
  --data grant_type=urn:ietf:params:oauth:grant-type:token-exchange \
  --data subject_token=eyJ<REDACTED>WGw \
  --data client_id=oidc-client-from-extIdp \
  --data client_secret=cb409d7b-7dc6-479d-84f8-45e8fc5189e8 \
  --data subject_token_type=urn:ietf:params:oauth:token-type:jwt \
  | jq -r '.access_token'`
3. int to ext from extIdp
`curl --request POST \
--url http://192.168.168.211:8081/auth/realms/ext-idp/protocol/openid-connect/token \
--header 'content-type: application/x-www-form-urlencoded' \
--header 'x-csrf-token: 1' \
--data grant_type=urn:ietf:params:oauth:grant-type:token-exchange \
--data subject_token=eyJ<REDACTED>Mg \
--data client_id=oidc-client-from-intIdp \
--data client_secret=3e5e4547-cd0d-49af-84cc-5c87981be7ef \
--data subject_token_type=urn:ietf:params:oauth:token-type:jwt`
##end's up with error_
`10:21:12,971 WARN  [org.keycloak.events] (default task-73) type=TOKEN_EXCHANGE_ERROR, realmId=ext-idp, clientId=oidc-client-from-intIdp, userId=null, ipAddress=172.17.0.1, error=federated_identity_account_exists, auth_method=token_exchange, grant_type=urn:ietf:params:oauth:grant-type:token-exchange, subject_issuer=http://192.168.168.211:8080/auth/realms/int-idp, validation_method=signature, client_auth_method=client-secret`

**Comment by diodfr** [comment]
> Hi, i tested this feature (https://www.keycloak.org/docs/latest/server_admin/#automatically-link-existing-first-login-flow) and this feature(https://www.keycloak.org/docs/latest/server_admin/#_disabling_automatic_user_creation) on KC 13.0.0, with token_exchange: i created two bi.directionally trusted IdP-s intIpd and extIdP(only he has users) then adjust on extIdP IdP First loging flow to custom authenticator flow from above examples, ,but autolink is not working!
> 
> 1. obtain ext JWT with
>    `curl -v --insecure --request POST \   --url http://192.168.168.211:8081/auth/realms/ext-idp/protocol/openid-connect/token \   --header 'content-type: application/x-www-form-urlencoded' \   --header 'x-csrf-token: 1' \   --data grant_type=password \   --data username=test \   --data password=test \   --data client_id=oidc-client-from-intIdp \   --data client_secret=3e5e4547-cd0d-49af-84cc-5c87981be7ef \   --data scope='openid email profile' \   | jq -r '.access_token'`
> 2. ext to int on intIdp
>    ` curl --request POST \   --url http://192.168.168.211:8080/auth/realms/int-idp/protocol/openid-connect/token \   --header 'content-type: application/x-www-form-urlencoded' \   --header 'x-csrf-token: 1' \   --data grant_type=urn:ietf:params:oauth:grant-type:token-exchange \   --data subject_token=eyJ<REDACTED>WGw \   --data client_id=oidc-client-from-extIdp \   --data client_secret=cb409d7b-7dc6-479d-84f8-45e8fc5189e8 \   --data subject_token_type=urn:ietf:params:oauth:token-type:jwt \   | jq -r '.access_token'`
> 3. int to ext from extIdp
>    `curl --request POST \ --url http://192.168.168.211:8081/auth/realms/ext-idp/protocol/openid-connect/token \ --header 'content-type: application/x-www-form-urlencoded' \ --header 'x-csrf-token: 1' \ --data grant_type=urn:ietf:params:oauth:grant-type:token-exchange \ --data subject_token=eyJ<REDACTED>Mg \ --data client_id=oidc-client-from-intIdp \ --data client_secret=3e5e4547-cd0d-49af-84cc-5c87981be7ef \ --data subject_token_type=urn:ietf:params:oauth:token-type:jwt`
> 
> end's up with error_
> `10:21:12,971 WARN [org.keycloak.events] (default task-73) type=TOKEN_EXCHANGE_ERROR, realmId=ext-idp, clientId=oidc-client-from-intIdp, userId=null, ipAddress=172.17.0.1, error=federated_identity_account_exists, auth_method=token_exchange, grant_type=urn:ietf:params:oauth:grant-type:token-exchange, subject_issuer=http://192.168.168.211:8080/auth/realms/int-idp, validation_method=signature, client_auth_method=client-secret`

I haven't been notify of your message. I will have a look asap.

**Pull request merged**